Source/Services/ReverseCalls/WrappedAsyncStreamWriter.cs

```diff
@@ -68,37 +77,81 @@ public WriteOptions WriteOptions
         /// <returns>A task that will complete when the message is sent.</returns>
         public async Task WriteAsync(TServerMessage message)
         {
+            _logger.WritingMessage(_requestId, typeof(TServerMessage));
             _cancellationToken.ThrowIfCancellationRequested();
+
             if (!_writeLock.Wait(0))
```

Comment:
aren't there async locks? I think there might be one in the Nito.AsyncEx packages I'm using in EventHorizon

Reply:
Hmmm, I might just don't know the API well enough :P

Reply:
In this case the `.Wait(0)` is a non-blocking attempt to take the lock. If it fails, it uses the lock asynchronously below.